### PR TITLE
Feat/get notification message

### PR DESCRIPTION
**Pull request by Judge40**
### refactor: extract template functionality

Extract template functionality from the EmailService in to a new
TemplateService, allowing reuse by other services in the future.

---

### feat: ability to view notification messages

Add a new endpoint at `/api/history/message/{id}` to allow the message to
be retrieved for a given notification ID.
Move the existing `/api/history/{id}` endpoint to
`/api/history/trainee/{id}` to avoid any confusion with the new
endpoint.

Update HistoryService to find the notification history and rebuild the
message from the given template name and variables.

Update the TemplateService to handle localization of `java.util.Date`
objects, which is the data type used when restoring a timestamp from the
database.

Update the email templates to include headers for formatting the

Refactor duplicated parameterized test method sources in to a new test
utility class.

Update Java to version 21.

TIS21-5124
TIS21-5266

### CODE REVIEW

File: src/test/java/uk/nhs/tis/trainee/notifications/service/HistoryServiceIntegrationTest.java [hunk 1]
```diff
@@ -155,4 +164,43 @@ void shouldSortFoundNotificationsBySentAtWhenMultipleFound() {
...
+  void shouldNotRebuildMessageWhenNotificationNotFound() {
+    Optional<String> message = service.rebuildMessage(ObjectId.get().toString());
+
+    assertThat("Unexpected message.", message, is(Optional.empty()));
+  }
+
+  @ParameterizedTest
+  @MethodSource(
+      "uk.nhs.tis.trainee.notifications.MethodArgumentUtil#getEmailTemplateTypeAndVersions")
+  void shouldRebuildEmailMessageWhenNotificationFound(NotificationType notificationType,
```
Comment (Judge40) [inline src/test/java/uk/nhs/tis/trainee/notifications/service/HistoryServiceIntegrationTest.java:178]: I'm not thrilled at this test, it does what it needs to but it feels awkward and brittle in the same way as some of the other tests that verify template content.
I'd like to look at the Thymeleaf testing library as a better way to test each template, but I'm at my boy scouting limits for this submission.

File: src/main/java/uk/nhs/tis/trainee/notifications/service/TemplateService.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+   *
+   * @param variables The template variables to localize.
+   * @return The variables with any timestamps localized.
+   */
+  private Map<String, Object> localizeTimestamps(Map<String, Object> variables) {
+    Map<String, Object> localizedTemplateVariables = new HashMap<>(variables);
+
+    for (Entry<String, Object> entry : localizedTemplateVariables.entrySet()) {
+      ZonedDateTime localized = switch (entry.getValue()) {
+        case Instant instant -> ZonedDateTime.ofInstant(instant, ZoneId.of(timezone));
```
Comment (ReubenRobertsHEE) [inline src/main/java/uk/nhs/tis/trainee/notifications/service/TemplateService.java:80]: I was thinking this pattern matching was a Java 21 thing, but I see it's actually 17 😀

Comment (Judge40) [reply src/main/java/uk/nhs/tis/trainee/notifications/service/TemplateService.java:80]: It was preview in 17, full release in 21 💪🏻

File: build.gradle [hunk 3]
```diff
@@ -10,9 +10,11 @@ plugins {
 }
 
 group = "uk.nhs.tis.trainee"
-version = "1.1.0"
+version = "1.3.0"
```
[the remark this answers is not in the document]
Comment (ReubenRobertsHEE) [reply build.gradle:13]: Technically a breaking change, with the history trainee API endpoint changing 😁 - v2??

**Review by ReubenRobertsHEE — APPROVED** [review]
Nice to get some of the data localisation stuff out of the Email Service as well. 🌮

File: build.gradle (hunk 3 above)
[the remark this answers is not in the document]
Comment (Judge40) [reply build.gradle:13]: In theory yes, but I'm hiding behind it never having been released/used yet 😇

**Pull request merged**